spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -316,34 +316,51 @@ struct ProductionVisitor
                     build_parse_stage1_logic();
 
                     // Call stage 2.
-                    std::vector<Expression> args = {state().data,   state().begin,      state().cur,  state().trim,
+                    auto args = [&]() {
+                        auto args =
+                            std::vector<Expression>{state().data,   state().begin,      state().cur,  state().trim,
                                                     state().lahead, state().lahead_end, state().error};
+                        if ( addl_param )
+                            args.push_back(builder::id(addl_param->id()));
 
-                    if ( addl_param )
-                        args.push_back(builder::id(addl_param->id()));
+                        return args;
+                    };
 
                     builder()->addLocal("filtered", hilti::builder::strong_reference(type::Stream()));
 
                     if ( unit ) {
                         pb->guardFeatureCode(*unit->id(), {"supports_filters"}, [&]() {
                             // If we have a filter attached, we initialize it and change to parse from its output.
+                            auto offset1 =
+                                builder()->addTmp("offset1",
+                                                  builder::memberCall(builder::begin(builder::deref(state().data)),
+                                                                      "offset", {}));
+
                             auto filtered = builder::assign(builder::id("filtered"),
                                                             builder::call("spicy_rt::filter_init",
                                                                           {state().self, state().data, state().cur}));
 
                             auto have_filter = builder()->addIf(filtered);
                             pushBuilder(have_filter);
 
-                            auto args2 = args;
-                            builder()->addLocal("filtered_data", type::ValueReference(type::Stream()),
-                                                builder::id("filtered"));
-                            args2[0] = builder::id("filtered_data");
-                            args2[1] = builder::begin(builder::deref(args2[0]));
-                            args2[2] = builder::deref(args2[0]);
-                            builder()->addExpression(builder::memberCall(state().self, id_stage2, args2));
+                            auto pstate = state();
+                            pstate.data = builder()->addTmp("filtered_data", type::ValueReference(type::Stream()),
+                                                            builder::id("filtered"));
+                            pstate.cur = builder::deref(pstate.data);
+                            pstate.begin = builder::begin(pstate.cur);
+                            // FIXME(bbannier): set `lahead` and `lahead_end` to refer to the filtered view.
```

Comment:
@rsmmr I am a little stuck what to pass here, in particular for `lahead_end`. Do you have an idea?